src/tsolvers/egraph/EnodeStore.cc

```diff
@@ -142,58 +180,45 @@ vec<std::pair<PTRef,ERef>> EnodeStore::constructTerm(PTRef tr) {
         addDistClass(tr);
     }
 
-    // Add both the pure and the negated terms
-    if (logic.isBooleanOperator(tr) || logic.isBoolAtom(tr) || logic.isTrue(tr) || logic.isFalse(tr)) {
-        PTRef tr_pure;
-        PTRef tr_neg;
-        lbool sgn;
-        logic.purify(tr, tr_pure, sgn);
-        tr_neg = logic.mkNot(tr_pure);
+    bool makeRecursiveDefinition = needsRecursiveDefinition(tr);
+
+    ERef sym = ERef_Nil;
+    ERef cdr = ERef_Nil;
+
+    const Pterm& tm = logic.getPterm(tr);
+    if (makeRecursiveDefinition) {
+        sym = addSymb(tm.symb());
+        for (int j = tm.size() - 1; j >= 0; j--) {
+            assert(termToERef.has(tm[j])); // The child was not inserted
+            ERef car = termToERef[tm[j]];
+#ifdef VERBOSE_EUF
+            ERef prev_cdr = cdr;
+            assert (operator[](car).getRoot() == car);
+            assert (operator[](cdr).getRoot() == cdr);
+#endif
+            cdr = addList(car, cdr);
+        }
+    } else {
+        sym = addAnonSymb(tr);
+        cdr = ERef_Nil;
+    }
+
+    ERef er = addTerm(sym, cdr, tr);
+    new_enodes.push({tr, er});
 
-        // If tr is a complex Boolean operator, do not model the full logic but cut here (the ERef
-        // will be treated as a UF constant with the anon name from the logic).  Otherwise, (tr is
-        // a pure Boolean atom or its negation), use the term ref from the logic.
-        ERef sym = logic.isBooleanOperator(tr_pure) ? addSymb(logic.getSym_anon()) : addSymb(logic.getSymRef(tr_pure));
+    if (logic.hasSortBool(tr)) {
+        // Add both the pure and the negated terms
```

Comment:
Now in this `if` you are only adding the negated term, right? The pure term has already been added at this point, no?

Reply:
Yes that's true.  I forgot to update the comment.